src/main/java/io/leonis/algieba/ai/SimpleBoid.java

```diff
@@ -0,0 +1,113 @@
+package io.leonis.algieba.ai;
+
+import io.leonis.algieba.Spatial;
+import io.leonis.algieba.spatial.Moving;
+import java.util.Set;
+import org.nd4j.linalg.api.ndarray.INDArray;
+import org.nd4j.linalg.factory.Nd4j;
+
+/**
+ * The Class SimpleBoid.
+ *
+ * This class contains the functionality for a simple boids algorithm. The implementation is based
+ * on the notes and pseudocode found <a href="http://www.kfish.org/boids/pseudocode.html">here</a>.
+ *
+ * @author Rimon Oz
+ */
+public final class SimpleBoid<O extends Spatial & Moving> implements Boid<O> {
+
+  @Override
+  public INDArray respond(
+      final O currentBoid,
+      final Set<? extends O> otherBoids,
+      final double avoidanceScale,
+      final double velocityRange,
+      final double centeringScale
+  ) {
+    return this.collisionAvoidance(currentBoid, otherBoids, avoidanceScale)
+        .add(this.velocityMatching(currentBoid, otherBoids, velocityRange))
+        .add(this.flockCentering(currentBoid, otherBoids, centeringScale));
+  }
+
+  /**
+   * Returns a velocity vector for a specific boid, given a set of other boids and scaling factor,
+   * such that the boid will avoid collisions.
+   *
+   * @param currentBoid   The boid to compute the velocity vector for.
+   * @param otherBoids    A set of {@link Boid}s.
+   * @param scalingFactor The tuning parameter for the collision avoidance rule. High values result
+   *                      in large vectors.
+   * @return A velocity vector representing the suggested velocity for the specified boid such that
+   * the specific boid steers away from other boids in the flock.
+   */
+  private <O extends Spatial & Moving> INDArray collisionAvoidance(
```

Comment:
Your type witness uses the same generic parameter as the class, but it's actually a different one. this is quite unclear.